pkg/operator/encryption/controllers/migrators/inprocess_processor.go

```diff
@@ -0,0 +1,170 @@
+package migrators
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/pager"
+	"k8s.io/klog"
+)
+
+const (
+	defaultConcurrency = 10
+)
+
+// WorkerFunc function that is executed by workers to process a single item
+type WorkerFunc func(*unstructured.Unstructured) error
+
+// ListProcessor represents a type that processes resources in parallel.
+// It retrieves resources from the server in batches and distributes among set of workers.
+type ListProcessor struct {
+	concurrency   int
+	workerFn      WorkerFunc
+	dynamicClient dynamic.Interface
+	ctx           context.Context
+}
+
+// NewListProcessor creates a new instance of ListProcessor
+func NewListProcessor(ctx context.Context, dynamicClient dynamic.Interface, workerFn WorkerFunc) *ListProcessor {
+	return &ListProcessor{
+		concurrency:   defaultConcurrency,
+		workerFn:      workerFn,
+		dynamicClient: dynamicClient,
+		ctx:           ctx,
+	}
+}
+
+// Run starts processing all the instance of the given GVR in batches.
+// Note that this operation block until all resources have been process, we can't get the next page or the context has been cancelled
+func (p *ListProcessor) Run(gvr schema.GroupVersionResource) error {
+	var errs []error
+	listPager := pager.New(pager.SimplePageFunc(func(opts metav1.ListOptions) (runtime.Object, error) {
+		for {
+			allResource, err := p.dynamicClient.Resource(gvr).List(opts)
+			if err != nil {
+				klog.Warningf("List of %v failed: %v", gvr, err)
+				if errors.IsResourceExpired(err) {
+					token, err := inconsistentContinueToken(err)
+					if err != nil {
+						return nil, err
+					}
+					opts.Continue = token
+					klog.V(4).Infof("Relisting %v after handling expired token", gvr)
+					continue
+				} else if retryable := canRetry(err); retryable == nil || *retryable == false {
+					return nil, err // not retryable or we don't know. Return error and controller will restart migration.
+				} else {
+					if seconds, delay := errors.SuggestsClientDelay(err); delay {
+						time.Sleep(time.Duration(seconds) * time.Second)
+					}
+					klog.V(4).Infof("Relisting %v after retryable error: %v", gvr, err)
+					continue
+				}
+			}
+
+			migrationStarted := time.Now()
+			klog.V(4).Infof("Migrating %d objects of %v", len(allResource.Items), gvr)
+			err = p.processList(allResource)
+			if err != nil {
+				errs = append(errs, err)
```

Comment:
I couldn't decide what to do, should we continue or should we stop and restart.

In general, we can't tolerate even a single error, so it could happen that the last item in the list could restart the whole migration.

Reply:
@sttts wdyt ?

Reply:
I think we have to stop over, i.e. return this error directly.